### PR TITLE
feat: Add ShortGPT pruning algorithm

### LINKED ISSUE #418: [FEATURE] ShortGPT pruning algorithm
## Is your feature request related to a problem? Please describe.

Current pruning algorithms in Pruna (e.g. torch_structured, torch_unstructured) focus on intra-layer compression — removing neurons, channels, or heads.

I came across a paper (e.g., ShortGPT: Layers in Large Language Models are More Redundant Than You Expect, Baichuan Inc) shows that many transformer layers contribute minimally to the model’s overall function, especially in pre-norm architectures like LLaMA or Baichuan, which is inter-layer compression.

Without layer-level pruning support, users can’t exploit this redundancy to reduce depth and latency while keeping accuracy.

## Describe the solution you'd like

Add a new pruning algorithm: **`shortgpt_layer_pruner`**, implementing the *ShortGPT* approach for **layer-level structured pruning** based on *Block Influence (BI)*.

### Overview

ShortGPT identifies and removes redundant transformer layers by measuring how much each layer actually transforms its input hidden states.
Layers whose inputs and outputs are nearly identical are considered low-impact and can be safely removed without significant performance loss.

### Core Algorithm

1. **Run calibration forward pass** on a small unlabeled dataset (e.g. PG19).
2. **Collect hidden states** for each transformer block: `h₀, h₁, …, h_L`.
3. **Compute Block Influence (BI)** for each layer *i*:

   ```python
   BI_i = 1 - mean(cosine_similarity(h_i, h_{i+1}))
   ```

   (Optionally support angular distance variant.)
4. **Rank layers by BI** (lower = more redundant).
5. **Remove lowest-BI layers** according to a configured ratio or count.
6. (Optional) **Post-train** lightweight fine-tuning or replace removed layers with small MLP adapters to recover minor performance loss.

### Integration with Pruna

* Implement as a subclass of `PrunaPruner`, e.g.

  ```python
  class ShortGPTPruner(PrunaPruner):
      algorithm_name = "shortgpt"
      ...
  ```
* Register under the `PRUNER` algorithm group.
* Compatible with quantizers like `torchao`, `half`, and `hqq` (since ShortGPT is orthogonal to quantization).
* Expected hyperparameters in `get_hyperparameters()`:

  * `num_layers_to_prune`: integer (default: 8)
  * `angular_metric`: boolean (default: False)
  * `calibration_samples`: integer (default: 64)
  * `pruning_ratio`: float (default: 0.25)

### Expected Benefits

* ~25% parameter and compute reduction with minimal (<10%) accuracy drop.
* No gradient or label requirements — runs purely from forward activations.
* Simpler than gradient-based or dependency-graph pruning.
* Extends easily to non-transformer models (RWKV, Mamba).
* Orthogonal to quantization, enabling compound compression (layer removal + 4-bit GPTQ).

---

## Additional context

Paper link: https://arxiv.org/abs/2403.03853

@sharpenb Let me know your thoughts on this. If you think this could be a useful addition, assign me and I can send in my PR.

### PR DESCRIPTION
**Pull request by sky-2002**
## Description
Implements the ShortGPT pruning algorithm.
Refer #418 

## Related Issue

Fixes #418 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Additional Notes
I am figuring out how to write tests for this, found the test framework a little tricky. Would appreciate some help there. I did not understand how the `run_full_integration` gets the dataset required to run tests, tokenizer etc.

Also, I found the `SmashConfigPrefixWrapper` confusing, so I tested with an alternative way:
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from pruna.algorithms.shortgpt import ShortGPT
from datasets import load_dataset
algo = ShortGPT()

llama_tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-1B", 
llama_model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.2-1B", torch_dtype="auto")

dataset = load_dataset("wikitext", "wikitext-2-raw-v1", split="train[:100]")

dl = DataLoader(dataset, batch_size=2, shuffle=True)

qmodel = algo._apply(llama_model, {
    "tokenizer": llama_tokenizer,
    "device": "cpu",
    "prune_ratio": 0.25,
    "angular": True,
    "train_dataloader": dl,
})

### CODE REVIEW

**Comment by sdiazlor** [comment]
Hi @sky-2002! Thank you again for the contribution.

- Have you checked the implementation after the latest updates?
- Are you currently encountering any trouble?
